### ISSUE TITLE
Implement interface implementations in GOOL

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1990]
The ODE solver for Java uses interfaces. More specifically, it requires writing an implementation for an existing interface. 

The concept to encode here is that of a class where all of the methods need to be overridden by sub-classes. The way that is achieved in Java and C# is through interfaces, whereas in Python and C++ you can do it with regular classes and multiple inheritance. So we should have a GOOL method that maps to the respective language's way of doing things.

And since the ODE solver only requires implementing an interface, not defining one, we can just support that for now (rather than also needed to support definition of interfaces).

Java also uses an anonymous inner class to implement an interface inline in a function body. There seem to be equivalents for this in the other languages too (`type` function in Python, anonymous class in C++, nested class definition in C#)

**Comment by smiths** [comment]
Sounds good @bmaclach.   I like the way that additions are being made to GOOL on an as needed basis.  This fits with the philosophy that we presented in the GOOL paper.  :-)

If we do later add support for defining an interface, my guess is that wouldn't be too bad.  It is like a class with an implementation, just without the implementation.  :-)

**Comment by JacquesCarette** [comment]
The concept of "implementing an interface" is quite important, so being able to do that makes a lot of sense.  Even if in some languages it doesn't generate 'fancy' code.

**Comment by bmaclach** [comment]
Looking closer at the second part of this issue, about the anonymous inner class, I've found that it's not actually achievable in all of the languages like I thought. It does seem possible in Python by using the `types` library and "assigning" an object's methods to new functions (see the answer by "gnud" to this question for more details: https://stackoverflow.com/questions/357997/does-python-have-something-like-anonymous-inner-classes-of-java). But I can't find a way to do it in either C# or C++ (the nested class/anonymous class I originally thought would work will not work because they don't have the function's local variables in scope).

So I went back to the drawing board of manual Java code in the case studies repo and found a way to refactor the NoPCM Java implementation so that it doesn't use an anonymous inner class. Instead, it just defines a regular, non-anonymous class, which seems to me like a fine solution. So I don't think I'll implement anything for anonymous inner classes -- what I've done in #2002 should be enough, and then I'll just update the `solveODE` code in GOOL to match the refactor I've done in the case studies repo, which will close this issue.
